Add polar coordinate `sf::Vector2<T>` constructor

ChrisThrasher opened the pull request:
## Description

This is the corollary to [`sf::Vector2<T>::angle()`](https://github.com/SFML/SFML/blob/master/include/SFML/System/Vector2.hpp#L121) which produces an angle from a vector. To avoid circular dependencies and to ensure we can construct vectors of all types from angles, I made this a new `sf::Vector2<T>` constructor overload instead of adding a new member function to `sf::Angle`. 

The motivation for this was simplifying how we currently create unit vectors from angles which looks like [`sf::Vector2f::UnitX.rotatedBy(rotation)`](https://github.com/ChrisThrasher/boids/blob/sfml-v3/src/boid.cpp#L23). I wouldn't expect many users to realize this is possible since at first glance it's not obvious what I'm trying to do here.

As an aside, this PR also removes the degree sign from Vector2.hpp which is an extended ASCII character. SFML appears to avoid extended ASCII characters in source files so I wanted to bring this file in line with the rest of the library.

## Tasks

* [x] Tested on Linux
* [x] Tested on Windows
* [x] Tested on macOS
* [x] Tested on iOS
* [x] Tested on Android

Bromeon commented:
Nice idea! I think the constructor should take angle **and length** though.
Basically the polar-coordinate equivalent `(r, phi)` to cartesian `(x, y)`.

Also, we should disallow construction from angles for non-float vector types.

ChrisThrasher commented:
> Nice idea! I think the constructor should take angle **and length** though. Basically the polar-coordinate equivalent `(r, phi)` to cartesian `(x, y)`.

Would you rather the length be an optional parameter or keep two constructors, one for unit vectors and one for vectors with a specified length?

> Also, we should disallow construction from angles for non-float vector types.

Good point. I'll fix that right now.

Bromeon commented:
> Would you rather the length be an optional parameter or keep two constructors, one for unit vectors and one for vectors with a specified length?

Personally I would not treat the `(1, phi)` construction as a special case, as we don't do that for any value of `x`, `y` or `phi` either. 

A general `(r, phi)` constructor should be a good trade-off regarding flexibility and API surface.

Bromeon commented on include/SFML/System/Vector2.hpp:
```diff
@@ -73,6 +73,14 @@ class Vector2
+    /// \param angle Angle
+    ///
+    ////////////////////////////////////////////////////////////
+    explicit Vector2(const Angle& angle);
```
I'd probably pass `Angle` by value and name it `phi` like here:
https://github.com/SFML/SFML/blob/c7327353b3eaa1ac62c13195a1424033121fb5de/include/SFML/System/Vector2.hpp#L140

(also applies once the constructor signature changes)

ChrisThrasher replied:
Weird, the implementation uses the name `angle`. I'll use `phi` if you want though.

https://github.com/SFML/SFML/blob/3ae85854e4674debe3b30ae2f6cf9d9926738789/include/SFML/System/Vector2.inl#L110

Bromeon replied:
Oh, that's an oversight, feel free to fix it. Ultimately it's not _that_ important 🙂

ChrisThrasher replied:
Might be worth some library-wide rename alongside changing many `const Angle&` params to `Angle`. For a later PR but I'll make a note to check for consistent `sf::Angle` use.

Bromeon replied:
I was mostly thinking about calling an angle `phi` in the context of polar coordinates (i.e. a vector) -- i.e. in cases where there is another coordinate `r` (radius). Not sure if that necessarily applies to the rest of the code base; so `angle` might still be fine elsewhere.

ChrisThrasher commented:
@Bromeon I changed that constructor to use a length/angle pair and added some new tests cases to make sure everything working as intended. I agree this is a better plan than two constructors and/or something involving optional parameters.

Bromeon commented on examples/tennis/Tennis.cpp:
```diff
@@ -200,8 +200,7 @@ int main()
             // Move the ball
-            float factor = ballSpeed * deltaTime;
-            ball.move({std::cos(ballAngle.asRadians()) * factor, std::sin(ballAngle.asRadians()) * factor});
+            ball.move({ballSpeed * deltaTime, ballAngle});
```
Nice one! This also shows the strength of a type-safe `sf::Angle` type, otherwise we would have needed something like `sf::Vector2f::fromPolar(r, phi)`.

Are people generally OK with this being a constructor, i.e. allowing short syntax?

Bromeon commented on test/System/Vector2.cpp:
```diff
@@ -37,6 +37,48 @@ TEST_CASE("sf::Vector2 class template - [system]")
+
+        SUBCASE("Length and angle constructor")
+        {
+            CHECK(sf::Vector2f(0, sf::degrees(0)) == sf::Vector2f(0, 0));
+            CHECK(sf::Vector2f(0, sf::degrees(180)) == sf::Vector2f(0, 0));
```
Could we maybe add some arbitrary values for `phi` here (not just half-circles), to see if it's still the zero vector?
Possibly we need `ApproxVec` for rounding errors.

ChrisThrasher replied:
Fixed

Bromeon commented on include/SFML/System/Vector2.hpp:
```diff
@@ -73,6 +73,15 @@ class Vector2
     template <typename U>
     constexpr explicit Vector2(const Vector2<U>& vector);
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Construct the vector from a length and angle
+    ///
+    /// \param length Length of vector
+    /// \param phi    Angle off X axis
+    ///
+    ////////////////////////////////////////////////////////////
+    Vector2(T length, Angle phi);
```
Do we want to allow negative values of `length`? (as it's currently handled in tests)

* If yes, we should probably rename the parameter, since it's no longer a length and it won't match the getter `length()`. Maybe `r`.
* Also, in that case we should document that it is allowed, and maybe even that `Vector2(r, phi)` is equivalent to `Vector2(-r, phi + 180°)`.
* Should we mention something about `Vector2(r, phi)` being equivalent to `Vector2(r, phi + n * 360°)`? 

Those things may sound obvious, but I think the important message is that the polar angle constructor is lossy, and it may not be able to retrieve the values later as they were provided in the constructor.

Bromeon replied:
Furthermore, maybe we should change the `\brief` part to:
```cpp
/// \brief Construct the vector from polar coordinates
```
while for the `(x, y)` constructor, we could say:
```cpp
/// \brief Construct the vector from cartesian coordinates
```

ChrisThrasher replied:
> Do we want to allow negative values of length? (as it's currently handled in tests)

In the Tennis example I modified, imagine if you wanted the ball to go backwards along a certain trajectory. In that case it's really helpful to use a negative radius/length instead of having extra logic to flip the angle. Besides, the math is perfectly well defined for all values of `r` and `phi`.

> If yes, we should probably rename the parameter, since it's no longer a length and it won't match the getter length(). Maybe r.

No problem. I'll adopt that r/phi terminology. I think it looks good and I like how it helps avoid confusion with `length()` and `angle()` which as you pointed out won't necessarily return the same values that were provided to this constructor. 

> Should we mention something about Vector2(r, phi) being equivalent to Vector2(r, phi + n * 360°)?

No harm in a little extra documentation :)

Bromeon commented on test/System/Vector2.cpp:
```diff
@@ -37,6 +37,48 @@ TEST_CASE("sf::Vector2 class template - [system]")
+            CHECK(sf::Vector2f(-1, sf::degrees(-270)) == ApproxVec(0, -1));
+            CHECK(sf::Vector2f(-1, sf::degrees(-360)) == ApproxVec(-1, 0));
+
+            CHECK(sf::Vector2f(3.14f, sf::degrees(0)) == sf::Vector2f(3.14f, 0));
+            CHECK(sf::Vector2f(3.14f, sf::degrees(45)) == ApproxVec(3.14f * std::sqrt(2.f) / 2.f, 3.14f * std::sqrt(2.f) / 2.f));
+            CHECK(sf::Vector2f(3.14f, sf::degrees(90)) == ApproxVec(0, 3.14f));
+            CHECK(sf::Vector2f(3.14f, sf::degrees(135)) == ApproxVec(-3.14f * std::sqrt(2.f) / 2.f, 3.14f * std::sqrt(2.f) / 2.f));
+            CHECK(sf::Vector2f(3.14f, sf::degrees(180)) == ApproxVec(-3.14f, 0));
+            CHECK(sf::Vector2f(3.14f, sf::degrees(270)) == ApproxVec(0, -3.14f));
+            CHECK(sf::Vector2f(3.14f, sf::degrees(360)) == ApproxVec(3.14f, 0));
+            CHECK(sf::Vector2f(3.14f, sf::degrees(-90)) == ApproxVec(0, -3.14f));
+            CHECK(sf::Vector2f(3.14f, sf::degrees(-180)) == ApproxVec(-3.14f, 0));
+            CHECK(sf::Vector2f(3.14f, sf::degrees(-270)) == ApproxVec(0, 3.14f));
+            CHECK(sf::Vector2f(3.14f, sf::degrees(-360)) == ApproxVec(3.14f, 0));
```
Is there a particular reason why you use "nearly Pi"? It doesn't seem like you make use of any of its properties, so I wonder if another arbitrary magic number might be less suggestive.

ChrisThrasher replied:
Haha yeah that was a bad idea. I'll change that.

ChrisThrasher commented:
Rebasing on master and fixing the new compiler warnings added by #2070 would require making changes that #2089 renders unnecessary so I'd prefer to block this PR until a decision can be reached on #2089.

ChrisThrasher commented:
@Bromeon Anything more you want to discuss about this?

Bromeon reviewed (approved):
Looks good to me!

eXpl0it3r commented:
SFMLBuildMaster: Build this please.

eXpl0it3r commented:
SFMLBuildMaster: Build this please.

ChrisThrasher commented:
Rebased after #2055 was merged just to be sure that there we no new conflicts.

eXpl0it3r commented:
Thanks for this addition. Lets see how it will be received/utilized. 🙂
